docs/config.rst

```diff
@@ -0,0 +1,180 @@
+  - "stash directory" for file-data of the snapshot
+
+NOTE: Currently the "list of other participants" comes from the
+"collective dircap" but in the Leif Design each user decides whom they
+wish to pull updates from. So, the "list of other participants" above
+might be "a readcap of a directory" (and we would determine the actual
+list at runtime by reading that capability).
+
+
+UI to Change the Config
+-----------------------
+
+Writing configuration files is hard, error-prone and not a great
+interface. It also makes it hard (or impossible) to change where or
+how configuration is stored. All configuration shall be changed by an
+HTTP API. There will also be CLI commands that are thin wrappers of
+the HTTP API. GUI front-ends could use the command-line or the HTTP
+API. Nothing except the HTTP API should be touching the actual stored
+configuration.
+
+(Of course we can't stop users from editing the config files
+themselves, but this should be discouraged).
+
+
+.. _current:
+
+Where and How To Store the Configuration
+----------------------------------------
+
+Somewhat mirroring Tahoe-LAFS's design, configuration shall all be
+found in a single directory. This directory can be specified via
+command-line option. (Should we have a default? Seems sensible, most
+OSes have a reasonably well-defined place for programs to store
+configuration data).
+
+All global configuration and state shall be stored in an SQLite3
+database. This database will have tables for global configuration and
+state.
+
+The "configuration directory" will look like this:
+
+- `confdir` (e.g. `~/.config/magic-folder`)
+  - `global.sqlite`
+  - `api_token`: 32 bytes of binary data
+  - `magic_folder.pid`: contains the PID of the currently-running process
+
+The `global.sqlite` database will have the following tables:
+
+- "version" with a single row containing "1"
+- "magic_folders"
+  - columns: `name`, `location`
+  - `location` is a local path for that folder's configuration
+- "config"
+  - columns: "api_endpoint", "tahoe_client_url"
+  - `api_endpoint` is a Twisted server string description
+  - `tahoe_client_url` is a URL to our Tahoe-LAFS client's Web-UI
+
+Each "magic folder" location is a directory containing the
+stash-directory and state for that magic-folder. It will look like
+this:
+
+- `<magic folder location>/`
+  - `README`: a file containing information about magic-folder
+  - `state.sqlite`: the state database
+  - `stash/`: the stash-directory containing LocalSnapshot content
+
+The `state.sqlite` for each magic-folder shall contain the following
+tables:
+
+- "version" (will always contain 1 row)
+  - column: "version"
+  - "version" is an int, currently `1`
+- "config" (will always contain 1 row)
+  - columns: "author_name", "author_private_key", "poll_interval"
+  - "author_name" is a string of unicode
+  - "author_private_key" is a 32-byte blob (a NaCl Signing key)
+  - "stash_path" is a local path where we stash data awaiting upload
```

Comment:
We could make this a path relative to the directory containing this database, making the whole config directory somewhat easily relocateable (or at least not imposing an additional barrier to relocation).

Reply:
otoh since it involves rewriting configuration we probably want some kind of `magic-folder move-folder-state` command rather than just letting people `mv` stuff, so the path could always be rewritten by that tool.